src/atlas/meshgenerator/detail/CubedSphereDualMeshGenerator.cc

```diff
@@ -0,0 +1,392 @@
+
+
+void CubedSphereDualMeshGenerator::configure_defaults() {
+    // This option sets number of partitions.
+    options.set( "nb_parts", mpi::size() );
+
+    // This option sets the part that will be generated.
+    options.set( "part", mpi::rank() );
+
+    // This options sets the number of halo elements around each partition.
+    options.set( "halo", 1 );
+
+    // This options sets the default partitioner.
+    options.set<std::string>( "partitioner", "cubedsphere" );
+}
+
+// -----------------------------------------------------------------------------
+
+void CubedSphereDualMeshGenerator::generate( const Grid& grid, Mesh& mesh ) const {
+    // Get partitioner type and number of partitions from config.
+    const idx_t nParts         = static_cast<idx_t>( options.get<size_t>( "nb_parts" ) );
+    const std::string partType = options.get<std::string>( "partitioner" );
+
+    auto partConfig = util::Config{};
+    partConfig.set( "type", partType );
+    partConfig.set( "partitions", nParts );
+
+    // Use lonlat instead of xy for non cubedsphere partitioner.
+    if ( partType != "cubedsphere" ) {
+        partConfig.set( "coordinates", "lonlat" );
+    }
+
+    // Set distribution.
+    const auto partitioner  = grid::Partitioner( partConfig );
+    const auto distribution = grid::Distribution( grid, partitioner );
+
+    generate( grid, distribution, mesh );
+}
+
+// -----------------------------------------------------------------------------
+
+void CubedSphereDualMeshGenerator::generate( const Grid& grid, const grid::Distribution& distribution, Mesh& mesh ) const {
+    // Check for correct grid and need for mesh
+    ATLAS_ASSERT( !mesh.generated() );
+
+    // Cast grid to cubed sphere grid.
+    const auto csGrid = CubedSphereGrid( grid );
+
+    // Check for successful cast.
+    if ( !csGrid ) {
+        throw_Exception(
+            "CubedSphereDualMeshGenerator can only work with a cubedsphere grid.", Here() );
+    }
+
+    // Check for correct grid stagger.
+    if ( csGrid.stagger() != "C" ) {
+        throw_Exception(
+            "CubedSphereDualMeshGenerator will only work with a cell-centroid grid.", Here() );
+    }
+
+    // Enforce compatible halo size.
+    if ( options.get<idx_t>( "halo" ) != 1 ) {
+        throw_Exception( "Halo size CubedSphereDualMeshGenerator is currently "
+                         "limited to 1.", Here() );
+    }
+
+    // Clone some grid properties.
+    setGrid( mesh, csGrid, distribution );
+
+    generate_mesh( csGrid, distribution, mesh );
+}
+
+// -----------------------------------------------------------------------------
+
+namespace  {
+
+// Helper function to copy fields.
+template<typename Value, int Rank>
+void copyField(const Field& sourceField, Field& targetField) {
+
+    // Assign source field values to target field.
+    array::make_view<Value, Rank>( targetField ).assign(
+                                array::make_view<Value, Rank>( sourceField ) );
+}
+
```

Comment:
@wdeconinck  I've added an assign overload to the ArrayView class so that this works. Hopefully that's a bit better.